source/common/config/watch_map.h

```diff
@@ -0,0 +1,118 @@
+#pragma once
+
+#include <set>
+#include <string>
+#include <utility>
+
+#include "envoy/config/subscription.h"
+
+#include "common/common/assert.h"
+#include "common/common/logger.h"
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/container/flat_hash_set.h"
+
+namespace Envoy {
+namespace Config {
+
+// Manages "watches" of xDS resources. Several xDS callers might ask for a subscription to the same
+// resource name "X". The xDS machinery must return to each their very own subscription to X.
+// The xDS machinery's "watch" concept accomplishes that, while avoiding parallel redundant xDS
+// requests for X. Each of those subscriptions is viewed as a "watch" on X, while behind the scenes
+// there is just a single real subscription to that resource name.
+//
+// This class maintains the watches<-->subscription mapping: it
+// 1) delivers updates to all interested watches, and
+// 2) tracks which resource names should be {added to,removed from} the subscription when the
+//    {first,last} watch on a resource name is {added,removed}.
+//
+// #1 is accomplished by WatchMap's implementation of the SubscriptionCallbacks interface.
+// This interface allows the xDS client to just throw each xDS update message it receives directly
+// into WatchMap::onConfigUpdate, rather than having to track the various watches' callbacks.
+//
+// The information for #2 is returned by updateWatchInterest(); the caller should use it to
+// update the subscription accordingly.
+//
+// A WatchMap is assumed to be dedicated to a single type_url type of resource (EDS, CDS, etc).
+class WatchMap : public SubscriptionCallbacks, public Logger::Loggable<Logger::Id::config> {
+public:
+  WatchMap() {}
+
+  // An opaque token given out to users of WatchMap, to identify a given watch.
+  using Token = uint64_t;
```

Comment:
I think an RAII pattern might work well here; return a Watcher object, put methods on it to update the watch, and then allow its destruction to signify the EOL of the watch. WDYT? This is a pattern that we use across Envoy and it generally is a lot cleaner than worrying about opaque token minting and resource tracking.

Reply:
That should work. I think I would add an `updateResources()` to the Watch interface. Then, the Watch implementation could encapsulate the token stuff, so that the SubscriptionImpl never needs to know about them.

It should also be possible to replace the tokens with pointers even within WatchMap, but I'm not comfortable with that for now. The tokens are a nice clean conceptual identifier thing to go into maps+sets. With pointers, they'd be doing double duty of acting as identifiers in maps+sets, but also providing direct access to the watches.

However, switching from tokens to pointers would skip some extra map lookups in every delta `onConfigUpdate()`, so it will be worth it - just maybe once we're satisfied that it's running cleanly, and we no longer need the easier-to-reason-about tokens.

Reply:
Conceptually, when do we need to map from token to anything if we encapsulate a Watches behavior entirely in the `Watch`?

Reply:
I don't think that much encapsulation is possible. The interesting behavior here - the batching of updates and the tracking of "does any watch care about resource X" - happens across all watches. There needs to be a single entity outside of the Watches handing that.

The tokens can be replaced by Watch pointers. I started with the tokens because it kept things clean to reason about; I think it's a bit harder to follow pointers being used as map keys when reading (or writing) this code for the first time. If you're ok with it, I could make the switch to pointers within this PR.

Reply:
Yeah, I think I would prefer pointers, and probably RAII if possible to avoid deregistering. The RAI I wrapper can be a distinct object which just does the deregister (and provides indirects to the underlying object for method invocation), this can support shared semantics as needed.

Reply:
Ok, will do! And I realized, there's already the typedef to the opaque Token type, so changing that typedef from a uint64 to a pointer shouldn't really affect readability. Great!

Reply:
I'm in the middle of this. Fully slick RAII is not going to work unless we're willing to have DeltaSubscriptionImpl's member variables need a particular destructor ordering. We can still have the DeltaSubscriptionImpl own the unique_ptr<Watch> if we have the destructor explicitly reset() it (with a comment on why it's necessary), which seems cleaner than relying on destructor ordering.

Reply:
Done.